crates/red_knot_python_semantic/src/types/mro.rs

```diff
@@ -0,0 +1,429 @@
+
+            // Fast path for a class that has only a single explicit base.
+            //
+            // This *could* theoretically be handled by the final branch below,
+            // but it's a common case (i.e., worth optimizing for),
+            // and the `c3_merge` function requires lots of allocations.
+            [single_base_node] => {
+                let single_base = ClassBase::try_from_node(
+                    db,
+                    single_base_node,
+                    class.definition(db),
+                    class_stmt_node.type_params.is_some(),
+                );
+                single_base.map_or_else(
+                    |invalid_base_ty| {
+                        let bases_info = Box::from([(0, invalid_base_ty)]);
+                        Err(MroErrorKind::InvalidBases(bases_info))
+                    },
+                    |single_base| {
+                        if let ClassBase::Class(class_base) = single_base {
+                            if class_is_cyclically_defined(db, class_base) {
+                                return Err(MroErrorKind::CyclicClassDefinition);
+                            }
+                        }
+                        let mro = std::iter::once(ClassBase::Class(class))
+                            .chain(single_base.mro(db))
+                            .collect();
+                        Ok(mro)
+                    },
+                )
+            }
+
+            // The class has multiple explicit bases.
+            //
+            // We'll fallback to a full implementation of the C3-merge algorithm to determine
+            // what MRO Python will give this class at runtime
+            // (if an MRO is indeed resolvable at all!)
+            multiple_bases => {
+                if class_is_cyclically_defined(db, class) {
+                    return Err(MroErrorKind::CyclicClassDefinition);
+                }
+
+                let definition = class.definition(db);
+                let has_type_params = class_stmt_node.type_params.is_some();
+                let mut valid_bases = vec![];
+                let mut invalid_bases = vec![];
+
+                for (i, base_node) in multiple_bases.iter().enumerate() {
+                    match ClassBase::try_from_node(db, base_node, definition, has_type_params) {
+                        Ok(valid_base) => valid_bases.push(valid_base),
+                        Err(invalid_base) => invalid_bases.push((i, invalid_base)),
+                    }
+                }
+
+                if !invalid_bases.is_empty() {
+                    return Err(MroErrorKind::InvalidBases(invalid_bases.into_boxed_slice()));
+                }
+
+                let mut seqs = vec![VecDeque::from([ClassBase::Class(class)])];
+                for base in &valid_bases {
+                    seqs.push(base.mro(db).collect());
+                }
+                seqs.push(valid_bases.iter().copied().collect());
```

Comment:
```suggestion
                seqs.push(valid_bases.into_iter().collect());
```

Reply:
That doesn't work because we need access to `valid_bases` in the closure passed to `.ok_or_else()` immediately below